src/Functions/punycode.cpp

```diff
@@ -0,0 +1,440 @@
+            ada::idna::utf8_to_utf32(value, value_length, value_utf32.data());
+
+            const bool ok = ada::idna::utf32_to_punycode(value_utf32, value_puny);
+            if (!ok)
+            {
+                if constexpr (error_handling == ErrorHandling::Throw)
+                {
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "'{}' cannot be converted to Punycode", std::string_view(value, value_length));
+                }
+                else
+                {
+                    value_puny.clear();
+                    col_res_null->getData()[row] = 1;
+                }
+            }
+
+            res_data.insert(value_puny.c_str(), value_puny.c_str() + value_puny.size() + 1);
+            res_offsets.push_back(res_data.size());
+
+            prev_offset = offsets[row];
+
+            value_utf32.clear();
+            value_puny.clear(); /// utf32_to_punycode() appends to its output string
+        }
+    }
+};
+
+
+template <ErrorHandling error_handling_>
+struct PunycodeDecodeImpl
+{
+    static constexpr auto error_handling = error_handling_;
+    static constexpr auto name = (error_handling == ErrorHandling::Null) ? "punycodeDecodeOrNull" : "punycodeDecode";
+
+    static void vector(
+        const ColumnString::Chars & data,
+        const ColumnString::Offsets & offsets,
+        ColumnString::Chars & res_data,
+        ColumnString::Offsets & res_offsets,
+        ColumnUInt8::MutablePtr & col_res_null)
+    {
+        const size_t rows = offsets.size();
+        res_data.reserve(data.size()); /// just a guess, assuming the input is all-ASCII
+        res_offsets.reserve(rows);
+        if constexpr (error_handling == ErrorHandling::Null)
+            col_res_null = ColumnUInt8::create(rows, 0);
+
+        size_t prev_offset = 0;
+        std::u32string value_utf32;
+        std::string value_utf8;
+        for (size_t row = 0; row < rows; ++row)
+        {
+            const char * value = reinterpret_cast<const char *>(&data[prev_offset]);
+            const size_t value_length = offsets[row] - prev_offset - 1;
+
+            const std::string_view value_punycode(value, value_length);
+            const bool ok = ada::idna::punycode_to_utf32(value_punycode, value_utf32);
+            if (!ok)
+            {
+                if constexpr (error_handling == ErrorHandling::Throw)
+                {
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "'{}' is not a valid Punycode-encoded string", value_punycode);
+                }
+                else
+                {
+                    value_utf32.clear();
+                    col_res_null->getData()[row] = 1;
+                }
+            }
+
+            const size_t utf8_length = ada::idna::utf8_length_from_utf32(value_utf32.data(), value_utf32.size());
+            value_utf8.resize(utf8_length);
+            ada::idna::utf32_to_utf8(value_utf32.data(), value_utf32.size(), value_utf8.data());
```

Comment:
Probably it's better to reassign `utf8_length` to the result of `ada::idna::utf32_to_utf8` here.